packages/ics721/src/execute.rs

```diff
@@ -0,0 +1,473 @@
+        let mut info = info;
+        info.sender = deps.api.addr_validate(&eyeball)?;
+        let cw721::Cw721ReceiveMsg {
+            token_id,
+            sender,
+            msg,
+        } = msg;
+        let res = self.receive_nft(deps, env, info, TokenId::new(token_id), sender, msg)?;
+        // override method key
+        Ok(res.add_attribute("method", "execute_receive_proxy_nft"))
+    }
+
+    fn get_class_data(&self, deps: &DepsMut, sender: &Addr) -> StdResult<ClassData> {
+        // cw721 v0.17 and higher holds ownership in the contract
+        let ownership: StdResult<Ownership<Addr>> = deps
+            .querier
+            .query_wasm_smart(sender, &cw721_base::msg::QueryMsg::Ownership::<Addr> {});
+        let owner = match ownership {
+            Ok(ownership) => ownership.owner.map(|a| a.to_string()),
+            Err(_) => {
+                // cw721 v0.16 and lower holds minter
+                let minter_response: cw721_base_016::msg::MinterResponse = deps
+                    .querier
+                    .query_wasm_smart(sender, &cw721_base_016::QueryMsg::Minter::<Empty> {})?;
+                deps.api.addr_validate(&minter_response.minter)?;
+                Some(minter_response.minter)
+            }
+        };
+        let contract_info = deps.querier.query_wasm_contract_info(sender)?;
+        let ContractInfoResponse { name, symbol } = deps
+            .querier
+            .query_wasm_smart(sender, &cw721_base::msg::QueryMsg::<Empty>::ContractInfo {})?;
+        let NumTokensResponse { count } = deps
+            .querier
+            .query_wasm_smart(sender, &cw721_base::msg::QueryMsg::<Empty>::NumTokens {})?;
+
+        Ok(ClassData {
+            owner,
+            contract_info,
+            name,
+            symbol,
+            num_tokens: count,
+        })
+    }
+
+    fn receive_nft(
+        &self,
+        deps: DepsMut,
+        env: Env,
+        info: MessageInfo,
+        token_id: TokenId,
+        sender: String,
+        msg: Binary,
+    ) -> Result<Response<T>, ContractError> {
+        let sender = deps.api.addr_validate(&sender)?;
+        let msg: IbcOutgoingMsg = from_binary(&msg)?;
+
+        let class = match NFT_CONTRACT_TO_CLASS_ID.may_load(deps.storage, info.sender.clone())? {
+            Some(class_id) => CLASS_ID_TO_CLASS.load(deps.storage, class_id)?,
+            // No class ID being present means that this is a local NFT
+            // that has never been sent out of this contract.
+            None => {
+                let class_data = self.get_class_data(&deps, &info.sender)?;
+                let class = Class {
+                    id: ClassId::new(info.sender.to_string()),
+                    // There is no collection-level uri nor data in the
+                    // cw721 specification so we set those values to
+                    // `None` for local, cw721 NFTs.
+                    uri: None,
+                    data: Some(to_binary(&class_data)?),
```

Comment:
collect ClassData when `receive_nft` is called. ClassData is pass for IBC message for transferring to another chain.

Reply:
Yeah that was the way we designed it.